internal/agent/client/client.go

```diff
@@ -72,6 +76,88 @@ func IsComposeAvailable() bool {
 	return false
 }
 
+type PullSecret struct {
+	// Absolute path to the pull secret
+	Path string
+	// Cleanup function for temporary files, or no-op
+	Cleanup func()
+}
+
+// ResolvePullSecret returns the image pull secret path, preferring inline spec
+// auth then falling back to on disk.  Cleanup removes tmp files generated from
+// inline spec if found and is otherwise a no-op.
+func ResolvePullSecret(
+	log *log.PrefixLogger,
+	rw fileio.ReadWriter,
+	desired *v1alpha1.DeviceSpec,
+	authPath string,
+) (*PullSecret, bool, error) {
+	specContent, found := authFromSpec(log, desired, authPath)
+	if found {
+		exists, err := rw.PathExists(authPath)
+		if err != nil {
+			return nil, false, err
+		}
+		if exists {
+			diskContent, err := rw.ReadFile(authPath)
+			if err != nil {
+				return nil, false, fmt.Errorf("reading existing auth file: %w", err)
+			}
+
+			if bytes.Equal(diskContent, specContent) {
+				log.Debugf("Using on-disk pull secret (identical to spec): %s", authPath)
+				return &PullSecret{Path: authPath, Cleanup: func() {}}, true, nil
+			}
+		}
+		path, cleanup, err := fileio.WriteTmpFile(rw, "os_auth_", "auth.json", specContent, 0600)
+		if err != nil {
+			return nil, false, fmt.Errorf("writing inline auth file: %w", err)
+		}
+		log.Debugf("Using inline auth from device spec")
+		return &PullSecret{Path: path, Cleanup: cleanup}, true, nil
+	}
+
+	exists, err := rw.PathExists(authPath)
+	if err != nil {
+		return nil, false, err
+	}
+	if exists {
+		log.Debugf("Using on-disk pull secret: %s", authPath)
+		return &PullSecret{Path: authPath, Cleanup: func() {}}, true, nil
+	}
+
+	return nil, false, nil
+}
+
+func authFromSpec(log *log.PrefixLogger, device *v1alpha1.DeviceSpec, authPath string) ([]byte, bool) {
+	if device.Config == nil {
+		return nil, false
+	}
+
+	for _, provider := range *device.Config {
+		spec, err := provider.AsInlineConfigProviderSpec()
```

Comment:
I know it won't actually cause issues as the json deserialization should behave properly here and non inline providers won't populate the `Inline` property, but when do we prefer checking the provider's type first?

Reply:
will cover in the followup

Reply:
done